test/hotspot/jtreg/vmTestbase/nsk/jdb/kill/kill001/kill001a.java

```diff
@@ -159,18 +157,27 @@ public void run() {
         synchronized (kill001a.lock) {}
 
         // Sleep during waitTime to give debugger a chance to kill debugee's thread.
-        // Note vthreads need a short sleep because they will never receive the kill,
-        // and therefore sleep the full time, resulting in a test timeout if too long.
         try {
-            Thread.currentThread().sleep(kill001a.vthreadMode ? 10000 : kill001a.waitTime);
```

Comment:
The PR description says:
`kill001 suspends all threads at a breakpoint and then does a "jdb kill" on each thread, which translate to ThreadReference.stop(), so this is expected to work now.`
As far as I see only main thread is suspended at "breakHere", all test threads wait in Thread.sleep().
I'd expect test threads to be unmounted in Thread.sleep(), so it's unclear how the test works in vthread mode

Reply:
You're right. I misunderstood what the test is doing. At the time of the `kill` commands, the threads are actually blocked here:

        synchronized (kill001a.lock) {}

Currently blocking in this manner pins the virtual thread to the carrier thread, so it is mounted. However, we are not yet in the try block at the time the async exception is thrown. After all the threads have had the async exception thrown at them, the main thread is resumed and the sychronized block is exited. The log shows that the exception is thrown on the sleep() line, which is the first statement after the synchronized, which just so happens to be within the try and the right thing happens.

However, even if I add another statement outside the synchronized and just before the try (which ends up being the line where the exception is thrown), it still passes. This is due to yet another bug. The exception is blowing us out of the thread, so we never hit the following block:

```
        if (!killed) {
            // Need to make sure the increment is atomic
            synchronized (kill001a.lock) {
                kill001a.notKilled++;
            }
        }
```

The debugger side checks for the proper notKilled value (it should be 0). So the test still passes even though it never confirmed that the correct async exception was thrown and caught. It just knows that the thread did not exit cleanly. Probably instead of tracking `notKilled`, the test should track `killed` and only increment when the correct exception is actually caught.

And one more thing, the following comment before the sleep() wrong:

        // Sleep during waitTime to give debugger a chance to kill debugee's thread.

I'm not so sure the sleep is even necessary since we block in the synchronized indefinitely until the async exception is thrown. I think just a simple statement at the start of the try block is good enough. Maybe a println(). But this is somewhat fragile because we need to make sure there is no code between the synchronized and the try, so we probably need to move the synchronized to within the try also.

Reply:
> The debugger side checks for the proper notKilled value (it should be 0). So the test still passes even though it never confirmed that the correct async exception was thrown and caught. It just knows that the thread did not exit cleanly. Probably instead of tracking `notKilled`, the test should track `killed` and only increment when the correct exception is actually caught.

I think this will make the test more reliable.

> I'm not so sure the sleep is even necessary since we block in the synchronized indefinitely until the async exception is thrown. I think just a simple statement at the start of the try block is good enough. Maybe a println(). But this is somewhat fragile because we need to make sure there is no code between the synchronized and the try, so we probably need to move the synchronized to within the try also.

looks like we just need to move synchronized (or some other code to force vthread to be mounted) to inside the try and remove Thread.sleep, no other statements are needed.

Reply:
> looks like we just need to move synchronized (or some other code to force vthread to be mounted) to inside the try and remove Thread.sleep, no other statements are needed.

That should work. At first I was unsure because the async exception is not actually thrown until executing the next statement after the synchronized. But there should be a branch at the end of the try block, and the exception should be thrown before it is executed. Hopefully the exception table covers that bytecode as being in the try block.

Reply:
This became a lot trickier to understand than I expected. The JVM does not have to throw the async exception at the next executed bytecode. The interpreter seems to do it at an invoke or after a goto (so the target of the goto is the instruction that will appear to have thrown the async exception). I think JIT'd code is similar, although maybe throws on a backwards branch and not after a forward branch. For this reason we need some sort of statement in place after the synchronized block to trigger the throwing of the exception before we leave the try block.

Another thing that messes up the test is that within the synchronized block, you can't have anything that will trigger the throw of the async exception. For example:

        try {
            synchronized (kill001a.lock) {
                kill001a.log.display("entered synchronized");
            }
            kill001a.log.display("exited synchronized");
        } catch (Throwable t) {

This messes up the test because now the exception will be thrown from within the synchronized block. This triggers a catch of the exception by an implicit catch clauses whose only purpose is to do the monitorexit and then rethrow the exception (which the explicit catch clause will then catch). The problem is that the test only expects one throw per thread, and now it gets 2. Although fixable on the debugger side, it is probably best to avoid.

I've pushed an update. Let me know if you think it is ok.

Reply:
I went to edit the PR description, but found it is not actually incorrect. It says "kill001 suspends all threads at a breakpoint". That is actually true. Although it is the main thread that hits the breakpoint, all threads are suspended. Since the additional threads are suspended and mounted, the TR.stop() can be done on them. If they are blocked (as they are) but not suspended, the TR.stop() would fail (for virtual threads). I'll try to clean it up a bit though. The description could be clearer.

Reply:
Yes, I realized this too. By default breakpoints have SUSPEND_ALL policy.